### PR TITLE
database: Validate (but do not migrate) schema on app startup

### LINKED ISSUE #25261: Decouple migrations: Simplify application startup sequence
Partially implements Step 2 in [RFC 469](https://docs.google.com/document/d/1_wqvNbEGhMzu45CHZNrLQYys7gl0AtVpLluGp4cbQKk).

Replace the current application startup checks (that only look at the Postgres version post-connection), as well as the code that blocks on schema updates, with a check that fails-fast if the schema is not the expected version (or above).

### PR DESCRIPTION
**Pull request by efritz**
This PR breaks the constructor methods in the `internal/database/connection/live` package to a `Migrate`/`Ensure` pair that will run migrations or return an error on an outdated schema (respectively).

All application startup sequences now use the `Ensure` variant, as we assume that the migrator has been run to success prior to this version of the application being deployed. _(Note that we bypass this in dev with a flag in the environment as we don't want to throw a wrench in developer iteration cycles.)_

This PR also modifies the startup of the `server` container to run the migration utility before starting up any application containers (but after starting up Postgres containers). See #28929, merged into this branch.

~**IMPORTANT**: Do not merge this until the migrator has been put in place in Cloud and managed instances, the k8s, docker-compose, and pure docker deployment instructions have been updated, and we have backwards compatibility checks running in CI. Once the migrator exists in deploy-sourcegraph, we should have the cluster-test.sh CI check pass and be able to merge this effort.~

Fixes #25261.

### CODE REVIEW

**Review by eseliger — CHANGES_REQUESTED** [review]
LGTM, requesting changes to block merge

**Review by mrnugget — APPROVED** [review]
screenshot_2021-12-10_16 06 55@2x

**Comment by efritz** [comment]
I've re-added the original connection methods and guarded the new behavior with an envvar in 43d8d37. This should be mergeable without breaking any existing behavior.

File: cmd/server/shared/shared.go [hunk 1]
```diff
@@ -194,17 +208,58 @@ func Main() {
...
+	}
+
+	log.Printf("Starting %s processes", name)
+	group.Go(func() error { return goreman.Start([]byte(strings.Join(procfile, "\n")), options) })
+}
+
+func runMigrator() {
+	log.Println("Starting migrator")
+
+	for _, schemaName := range []string{"frontend", "codeintel"} {
```
Comment (coury-clark) [inline cmd/server/shared/shared.go:254]: Am I correct in thinking `codeinsights-db` will need to be added once we've migrated from Timescale? If so, what happens in the meantime, can we still run migrations on it?

Comment (efritz) [reply cmd/server/shared/shared.go:254]: This is in the `server` image. Will insights be embedded into the single docker image deployment?

Comment (coury-clark) [reply cmd/server/shared/shared.go:254]: ah - didn't realize that. No, I don't think so!

**Review by coury-clark — APPROVED** [review]

**Comment by efritz** [comment]
Main dry run: https://buildkite.com/sourcegraph/sourcegraph/builds/121737

**Review by eseliger — APPROVED** [review]

**Pull request merged**